cardano-db-sync/src/Cardano/DbSync/Epoch.hs

```diff
@@ -84,75 +96,129 @@ epochInsert trce (BlockDetails cblk details) = do
-      insertBlock trce details
+      insertEpoch trce cache details
 
 -- -------------------------------------------------------------------------------------------------
 
-insertBlock ::
+insertEpoch ::
   Trace IO Text ->
+  Cache ->
   SlotDetails ->
   ReaderT SqlBackend (LoggingT IO) (Either SyncNodeError ())
-insertBlock trce details = do
-  mLatestCachedEpoch <- liftIO $ readIORef latestCachedEpochVar
-  let lastCachedEpoch = fromMaybe 0 mLatestCachedEpoch
-      epochNum = unEpochNo (sdEpochNo details)
+insertEpoch trce cache details = do
+  -- read the chache Epoch
+  latestCachedEpoch <- liftIO $ readCacheEpoch cache
+  let maybeEpoch = ceEpoch latestCachedEpoch
+
+  -- if there isn't cache epock number default it to 0
+  let lastCachedEpochNo = maybe 0 epochNo maybeEpoch
+      slotEpochNum = unEpochNo (sdEpochNo details)
 
   -- These cases are listed from the least likey to occur to the most
   -- likley to keep the logic sane.
-
   if
-      | epochNum > 0 && isNothing mLatestCachedEpoch ->
-          updateEpochNum 0 trce
-      | epochNum >= lastCachedEpoch + 2 ->
-          updateEpochNum (lastCachedEpoch + 1) trce
+      | slotEpochNum > 0 && isNothing maybeEpoch ->
+        updateEpochNum cache 0 trce
+      | slotEpochNum >= lastCachedEpochNo + 2 ->
+        updateEpochNum cache (lastCachedEpochNo + 1) trce
       | getSyncStatus details == SyncFollowing ->
-          -- Following the chain very closely.
-          updateEpochNum epochNum trce
+        updateEpochNum cache slotEpochNum trce
       | otherwise ->
-          pure $ Right ()
+        pure $ Right ()
 
 -- -------------------------------------------------------------------------------------------------
 
-{-# NOINLINE latestCachedEpochVar #-}
-latestCachedEpochVar :: IORef (Maybe Word64)
-latestCachedEpochVar = unsafePerformIO $ newIORef Nothing -- Gets updated later.
+updateEpochNum ::
+  (MonadBaseControl IO m, MonadIO m) =>
+  Cache ->
+  Word64 ->
+  Trace IO Text ->
+  ReaderT SqlBackend m (Either SyncNodeError ())
+updateEpochNum cache slotEpochNum trce = do
+  mid <- queryForEpochId slotEpochNum
+  maybe
+    (insertEpochDB cache trce slotEpochNum)
+    (updateEpochDB cache slotEpochNum)
+    mid
+
+-- expensive calculation try using cache to minimise query
+updateEpochDB ::
+  (MonadBaseControl IO m, MonadIO m) =>
+  Cache ->
+  Word64 ->
+  EpochId ->
+  ReaderT SqlBackend m (Either SyncNodeError ())
+updateEpochDB cache slotEpochNum epochId = do
+  cachedEpoch <- liftIO $ readCacheEpoch cache
+  case ceEpoch cachedEpoch of
+    -- only call queryCalc if we don't already have an epoch in cache
+    Nothing -> queryCalc
+    Just _ ->
+    -- Just cacheEp ->
+      case ceLastKnownBlock cachedEpoch of
+        Nothing -> queryCalc
+        Just _ ->
+        -- Just lastKnowBlockId ->
+          -- if we have both a block and epoch then lets use them
+          -- to do a new less expensive query of updating the epoch
+          queryCalc
```

Comment:
Might need some pointers in regards to the DB query that can utilise the last known block in cache 🤔

Reply:
from a top level I think the new query should:

1.  use last stored block in cache and retrieve all blocks that came after it (either by ID or time stamp)
2. sum the needed parameters of the blocks.
3. merger the result to the exiting epoch both in cache and DB.